pyanaconda/modules/boss/module_manager.py

```diff
@@ -0,0 +1,122 @@
+#
+# module_manager.py: Anaconda DBUS module management
+#
+# Copyright (C) 2017  Red Hat, Inc.  All rights reserved.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+from pyanaconda.dbus import dbus_constants
+
+from pyanaconda.anaconda_loggers import get_module_logger
+log = get_module_logger(__name__)
+
+
+class ModuleManager(object):
+
+    def __init__(self, bus):
+        self._bus = bus
+        self._dbus = self._bus.get('org.freedesktop.DBus', '/org/freedesktop/DBus')
+        self._started_module_services = []
+        self._failed_module_services = []
+        self._addon_module_services = []
+
+    @property
+    def addon_module_services(self):
+        return self._addon_module_services
+
+    @property
+    def expected_module_services(self):
+        expected_modules = []
+        # NOTE: we can just initialize expected_modules with the constant,
+        # bacuse the extend afterwards will then *modify* the constant! :P
+        expected_modules.extend(dbus_constants.ANACONDA_MODULES)
+        expected_modules.extend(self.addon_module_services)
+        return expected_modules
+
+    @property
+    def running_module_services(self):
+        # For our purpose, just this
+        return self._started_module_services
+
+    @property
+    def modules_starting_finished(self):
+        return set(self._started_module_services + self._failed_module_services) == set(self.expected_module_services)
+
+    def _start_service_cb(self, service, returned=None, error=None):
+        """Callback for dbus.StartServiceByName."""
+        if error:
+            log.debug("%s error: %s", service, error)
+            self._failed_module_services.append(service)
+        elif returned:
+            if returned == dbus_constants.DBUS_START_REPLY_SUCCESS:
+                log.debug("%s started successfully, returned: %s)", service, returned)
+                self._started_module_services.append(service)
+            else:
+                log.debug("%s failed to start, returned: %s", service, returned)
+                self._failed_module_services.append(service)
+        else:
+            log.error("%s failed to start without even returning anything", service)
+
+        self.check_modules_started()
```

Comment:
Personally I don't like this whole construction:

```
if:
elif:
   if:
   else:
else:
```
It looks like a puzzle to me to find something in there. Could it be more flat / more logically separated?

Also this `self.check_modules_started()` doesn't give sense to me. It should be started outside of this function. This is basically a giant side effect of this whole method.

Reply:
Should be better now.